ast/src/parsed/display.rs

```diff
@@ -574,7 +621,7 @@ impl<Ref: Display> Display for Expression<Ref> {
             Expression::Tuple(items) => write!(f, "({})", format_list(items)),
             Expression::LambdaExpression(lambda) => write!(f, "{}", lambda),
             Expression::ArrayLiteral(array) => write!(f, "{array}"),
-            Expression::BinaryOperation(left, op, right) => write!(f, "({left} {op} {right})"),
+            Expression::BinaryOperation(_, _, _) => self.format_binary_operation(f),
```

Comment:
I think we need to put parentheses here, because we do not know the outer precedence. You can try testing `-(x + y)`.

Reply:
Ok, I think the other way to fix this is to implement the precedence-aware formatting for unary operation and lambda expression.

Reply:
It might be sufficient to implement the `precedence()` function for them.